sys/stdio_nimble/stdio_nimble.c

```diff
@@ -0,0 +1,399 @@
+            char str[40];
+            sprintf(str, "rc = %d; abort sending on error\n", rc);
+            debug(str, strlen(str));
+        }
+
+        /* abort sending on error */
+        _sending_in_progress = false;
+    }
+
+    mutex_unlock(&_send_mutex);
+}
+
+static void _clear_ringbuffer(void)
+{
+    ringbuffer_init(&_rb_stdin, _rb_stdin_mem, CONFIG_STDIO_NIMBLE_STDIN_BUFSIZE);
+
+    mutex_lock(&_rb_stdout_mutex);
+    ringbuffer_init(&_rb_stdout, _rb_stdout_mem, CONFIG_STDIO_NIMBLE_STDOUT_BUFSIZE);
+    mutex_unlock(&_rb_stdout_mutex);
+}
+
+static int _gap_event_cb(struct ble_gap_event *event, void *arg)
+{
+    (void) arg;
+    char* d;
+
+    switch (event->type) {
+
+        case BLE_GAP_EVENT_CONNECT:
+            if (event->connect.status == 0) {
+                _clear_ringbuffer();
+                _conn_handle = event->connect.conn_handle;
+                ble_att_set_preferred_mtu(MTU_MAX_NIMBLE);
+                ble_gattc_exchange_mtu(_conn_handle, NULL, NULL);
+            }
+            else {
+                _send_stdio = false;
+            }
+            break;
+
+        case BLE_GAP_EVENT_DISCONNECT:
+            _send_stdio = false;
+            break;
+
+        case BLE_GAP_EVENT_SUBSCRIBE:
+            if (event->subscribe.attr_handle == _val_handle_stdout) {
+                if (event->subscribe.cur_indicate == 1) {
+                    _send_stdio = true;
+                }
+                else {
+                    _send_stdio = false;
+                }
+            }
+            break;
+
+        case BLE_GAP_EVENT_NOTIFY_TX:
+            if (event->notify_tx.indication == 1) {
+                if (event->notify_tx.status == BLE_HS_EDONE) {
+                    d = "event->notify_tx.status == BLE_HS_EDONE\n"; debug(d, strlen(d));
+                    _send_from_ringbuffer(false);
+                }
+                else if (event->notify_tx.status != 0) {
+                    char str[40];
+                    sprintf(str, "event->notify_tx.status = %d;\n", event->notify_tx.status);
+                    debug(str, strlen(str));
+
+                    _send_stdio = false;
+                }
+            }
+            break;
+
+        case BLE_GAP_EVENT_MTU:
+            _current_max_mtu_payload = event->mtu.value - PDU_HEADERS_SIZE;
+
+            char str[40];
+            sprintf(str, "_current_max_mtu_payload = %d\n", _current_max_mtu_payload);
+            debug(str, strlen(str));
+
+            break;
+    }
+
+    return 0;
+}
+
+static int gatt_svr_chr_access_stdin(
+        uint16_t conn_handle, uint16_t attr_handle,
+        struct ble_gatt_access_ctxt *ctxt, void *arg)
+{
+
+    (void) conn_handle;
+    (void) attr_handle;
+    (void) arg;
+
+    uint16_t om_len = OS_MBUF_PKTLEN(ctxt->om);
+
+    /* read sent data */
+    int rc = ble_hs_mbuf_to_flat(ctxt->om, _stdin_read_buf,
+            sizeof(_stdin_read_buf), &om_len);
+
+    ringbuffer_add(&_rb_stdin, _stdin_read_buf, om_len);
```

Comment:
Maybe `isrpipe` could be used here?

Reply:
If its unneeded then ignore.

Reply:
Addition to buffer was replaced by a call to `isrpipe_write`. This is a new function introduced by 25e8280c21d38d21788007b79a10b23cdb79c36a. I'll raise a PR for this.